### DISCUSSION TITLE
PlanarControls upgrades

### DISCUSSION THREAD

**Discussion by valentinMachado** [iTowns/itowns#2075] [Ideas]
In [ud-viz](https://github.com/VCityTeam/UD-Viz) our widget give the possibility to travel to a box3 which is useful to focus some feature of the batchtable for example. In order to contribute them to itowns we would need to have a similar function in order to not write this code multiple time.

Could be a method (initiateTravelToBox3 maybe) with the same signature as initiateTravel but instead of a target position this is a box 3, there would be some math to compute the target camera position based on the box 3 then it could call the initiateTravel method. Or maybe the math code belong to CameraUtil and the user write a combination of CameraUtil math method + initiateTravel of PlanarControls ?

What do you think ?

Another thing is that the initiateTravel does not return a Promise resolving when the travel is actually done, forcing itowns user to write something not quite elegant like this:

```
const promiseInitiateTravel(...args) {
return new Promise((resolve) => {
initiateTravel(...args)
setTimeout(resolve, timeOfTheTravel)
}};
```

Since initiateTravel do an initialization and is not actually updating camera state (do in handleTravel) the only way (with this code structure as far as I know) would be to create a reference `this.travelResolve` set in initiateTravel and call in handleTravel to allow the user to write something like:

```
initiateTravel().then(()=>//stuff) 
```
 Which can be nice for the user to write good asynchronous code but also kind of confusing since the initialization of the travel is not asynchronous (the travel is)

so maybe we could just create a method `travel`
doing the initiateTravel + set this.travelResolve and returning a Promise

What do you think ?

**Comment by jailln** [comment]
I agree that it would be a good idea to have a method to travel to a `Box3`. I add a same need and started a [PR on this matter](https://github.com/iTowns/itowns/pull/1904). I think it should be implemented in `CameraUtils` though instead of in the controls. I started to implement a method to get the `CameraTransformOptions` from a `Box3` so one can then do `CameraUtils.animateCameraToLookAtTarget()` to apply these `CameraTransformOptions` (we could add a method that calls both methods to improve the API so one can do something like `CameraUtils.animateCameraToLookAtBox3()` directly). I never took the time to finish this work ^^  Would it satisfy your need?

**Reply by valentinMachado** [reply]
>Yes sounds good, however when you said that there was a function `animate...` in `CameraUtils` I was like what ?!

>Why PlanarControls is not using CameraUtils animation stuff ?
>Why using tween to `update` the animation ?
>looks like [tween is updated in the main loop](https://github.com/iTowns/itowns/blob/90ada88f35ceb7f32c407085fa6db7f9554adde6/utils/debug/TileDebug.js#L205) but for debug tools
>so how does it works when you are not using the debug tools ? are you suppose to add yourself tween update as a frame requester ? 

=> I did not find this [one](https://github.com/iTowns/itowns/blob/b39edf81d97c9fca320c751fc3df4f56fd8d2a64/src/Utils/CameraUtils.js#L271) sorry, I spoke before checking there

Is there a particular reason to not use tween in PlanarControls ?
I 'll take a look at the PR.

**Reply by alexLuky** [reply]
I've found very cool CameraControl lib and it would be great to integrate it in iTowns. It has all usefull control states such as zoom into point, rotate around a mouse cursor point, truck and many other. Also we have a lot of cases when we need to hide the globe and navigate in the space. Current controls can't work with hidden globe.
[https://github.com/yomotsu/camera-controls](https://github.com/yomotsu/camera-controls)

I've tried to integrate it in itowns... but I couldn't.

**Reply by jailln** [reply]
@valentinMachado there is some cleaning to do in itowns controls and camerautils :wastebasket: See issues #1972 and #945 . Feel free to add to them.

@alexLuky I didn't know about this library. A first step would indeed to integrate it in itowns to test it out. It could at least be integrated by itowns users in their apps using itowns if they need it. To decide if we want to integrate it in itowns we should check it out and see what new features it would bring, check if no features are lost, assert that we want to delegate this part of the code to another lib and verify that the lib is sustainable

**Reply by alexLuky** [reply]
@jailln 
Please note that there are cases when we need to navigate under the globe surface and current realization tied to the globe mesh and as soon as you fall below the surface navigation totally breaks. Also there is need to look at mesh without the globe. This controls are useful because they use their own plane for raycast and navigation basis for space travelling.

**Reply by jailln** [reply]
Yes, #2077 is a first step towards seeing underneath the globe but it does not solve underground navigation yet. To visualize stuff without the globe you can also use the `PlanarView` or `View` for now
